src/core/public/keyboard_shortcut/keyboard_shortcut_service.ts

```diff
@@ -0,0 +1,129 @@
+ * Any modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+import { KeyboardShortcutSetup, KeyboardShortcutStart, ShortcutDefinition } from './types';
+
+/**
+ * @internal
+ * @experimental
+ */
+export class KeyboardShortcutService {
+  private shortcuts = new Map<string, ShortcutDefinition[]>();
+
+  public setup(): KeyboardShortcutSetup {
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+    };
+  }
+
+  public start(): KeyboardShortcutStart {
+    this.startEventListener();
+
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+      unregister: (id) => this.unregister(id),
+    };
+  }
+
+  public stop() {
+    this.stopEventListener();
+    this.shortcuts.clear();
+  }
+
+  private normalizeKeyboardShortcutString = (str: string): string => str.toLowerCase();
+
+  private getNamespacedIdForKeyboardShortcut = (
+    shortcut: Pick<ShortcutDefinition, 'id' | 'pluginId'>
+  ) =>
+    `${this.normalizeKeyboardShortcutString(shortcut.id)}.${this.normalizeKeyboardShortcutString(
+      shortcut.pluginId
+    )}`;
+
+  private getEventKeyString = (event: KeyboardEvent): string => {
+    return `${event.ctrlKey ? 'ctrl+' : ''}${event.altKey ? 'alt+' : ''}${
+      event.shiftKey ? 'shift+' : ''
+    }${event.metaKey ? 'cmd+' : ''}${this.normalizeKeyboardShortcutString(event.key)}`;
+  };
+
+  private register(shortcuts: ShortcutDefinition[]): void {
+    shortcuts.forEach((shortcut) => {
+      const key = this.normalizeKeyboardShortcutString(shortcut.keys);
+      const existingShortcuts = this.shortcuts.get(key) || [];
+      existingShortcuts.push(shortcut);
+      this.shortcuts.set(key, existingShortcuts);
+    });
+  }
+
+  private unregister(fullId: string): void {
+    for (const [key, shortcuts] of this.shortcuts.entries()) {
+      const filteredShortcuts = shortcuts.filter(
+        (shortcut) => this.getNamespacedIdForKeyboardShortcut(shortcut) !== fullId
+      );
+
+      if (filteredShortcuts.length === 0) {
+        this.shortcuts.delete(key);
+      } else if (filteredShortcuts.length !== shortcuts.length) {
+        this.shortcuts.set(key, filteredShortcuts);
+      }
+    }
+  }
+
+  private isHTMLElement(target: EventTarget | null): target is HTMLElement {
+    return target !== null && 'tagName' in target;
+  }
+
+  private shouldIgnoreKeyboardEventForTarget(target: EventTarget | null): boolean {
+    if (!this.isHTMLElement(target)) return false;
+
+    const tagName = target.tagName;
+
+    if (tagName === 'INPUT' || tagName === 'TEXTAREA') {
+      return true;
+    }
+
+    return false;
+  }
+
+  private handleKeyboardEvent = (event: KeyboardEvent): void => {
+    if (this.shouldIgnoreKeyboardEventForTarget(event.target)) {
+      return;
+    }
+
+    const eventKeyString = this.getEventKeyString(event);
+    const shortcuts = this.shortcuts.get(eventKeyString);
+
+    if (shortcuts && shortcuts.length > 0) {
+      event.preventDefault();
```

Comment:
add a comment that we want to prevent browser specific keybindings if they conflict